### ISSUE TITLE
prefer-destructuring should be smarter

### ISSUE THREAD

**Issue by iugo** [eslint/eslint#9666]
**Tell us about your environment**

* **ESLint Version:** 4.12.0
* **Node Version:** v8.9.1
* **npm Version:** 5.5.1

**What parser (default, Babel-ESLint, etc.) are you using?**
babel-eslint

**Please show your full configuration:**

<details>
<summary>Configuration</summary>

```js

{
  "parser": "babel-eslint",
  "env": {
    "browser": true
  },
  "parserOptions": {
    "ecmaVersion": 2017,
    "sourceType": "module",
    "ecmaFeatures": {
      "modules": true,
      "jsx": true
    }
  },
  "plugins": [
    "react"
  ],
  "extends": [
    "eslint:recommended",
    "plugin:react/recommended",
    "airbnb"
  ],
  "globals": {
    "fetch": false
  },
  "rules": {
    "arrow-parens": ["error", "as-needed"],
    "camelcase": "off",
    "key-spacing": ["error", { "mode": "minimum" }],
    "linebreak-style": ["error", "unix"],
    "max-len": 0,
    "no-underscore-dangle":0,
    "no-param-reassign": ["error", { "props": true }],
    "no-unused-vars":0,
    "prefer-const":"off",
    "no-plusplus":"off",
    "no-unused-expressions": ["error", {
      "allowShortCircuit": true
    }],
    "no-use-before-define": ["error", {
      "functions": false
    }],
    "react/prop-types": "off",
    "react/forbid-prop-types":"off",
    "react/jsx-filename-extension": [1, { "extensions": [".js", ".jsx"] }],
    "import/extensions": ["error", {
      "jsx": "always"
    }],
    "jsx-a11y/click-events-have-key-events": "off",
    "react/jsx-closing-tag-location": "off", // https://github.com/yannickcr/eslint-plugin-react/issues/1466
    "import/no-named-as-default":0
  }
}

```

</details>

**What did you do? Please include the actual source code causing the issue, as well as the command that you used to run ESLint.**

```js
  const num = Math.random();
  const pack = { word: 'Bye' };
  let word = 'Hi';
  if (num < 0.5) {
    word = pack.word; // the error line
  }
  console.log(word);
```

```bash

```

**What did you expect to happen?**

`word = pack.word` is correct code.

**What actually happened? Please include the actual, raw output from ESLint.**

`word = pack.word` is incorrect code.

**Comment by not-an-aardvark** [comment]
Closing as a duplicate of https://github.com/eslint/eslint/issues/9601
